ocaml/typing/env.ml

```diff
@@ -3153,9 +3220,14 @@ and lookup_all_args ~errors ~use ~loc lid0 env =
 and lookup_apply ~errors ~use ~loc lid0 env =
   let f0_lid, args0 = lookup_all_args ~errors ~use ~loc lid0 env in
   let args_for_errors = List.map (fun (_,p,mty) -> (p,mty)) args0 in
-  let f0_path, f0_comp =
+  let f0_path, f0_locks, f0_comp =
     lookup_module_components ~errors ~use ~loc f0_lid env
   in
+  if use then begin
+    let f0_mode = Mode.Value.legacy |> Mode.Value.disallow_right in
+    let f0_vmode = walk_locks ~errors ~loc ~env ~item:Module ~lid:f0_lid f0_mode None f0_locks in
+    submode_legacy ~errors ~loc ~env ~item:Module ~lid:f0_lid f0_vmode
```

Comment:
I don't think we should require that a functor be legacy to be applied. As with ordinary functions, we should allow functors to be applied at any mode and ensure at their definitions that this will be safe.

Reply:
functors can be partially applied, meaning there are some mode constraints between the partial parameter, the functor, and the resulted module. That sounds like some efforts required out of the scope this PR. If we just assume all modules are legacy whatsoever, then we don't need to worry about those.

Reply:
Rather than having to work out the soundness of a different set of typing rules for functors, I think it will be easier to use the same typing rules as for functions. Note that the "close over" morphism, applied to the legacy mode produces the legacy mode. This means that restricting all functor parameters to legacy and all functor returns to legacy is sound by our normal rules, including correctly handling partial application, so there is no need to add any mode constraints.

Reply:
I assume that you meant something like:
```ocaml
let foo @ portable =
  fun () -> 
    let module X = struct ... end in
    let module Z = F(X) in
    ...
```
Just as with ordinary functions, this should be rejected because `F` is not `portable` not because of anything to do with `contended`.